crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs

```diff
@@ -136,3 +196,226 @@ fn create_stores_for_v5(db: &IdbDatabase) -> Result<(), DomException> {
 
     Ok(())
 }
+
+fn create_stores_for_v6(db: &IdbDatabase) -> Result<(), DomException> {
+    // We want to change the shape of the inbound group sessions store. To do so, we
+    // first need to build a new store, then copy all the data over.
+    //
+    // But copying the data needs to happen outside the database upgrade process
+    // (because it needs async calls). So, here we create a new store for
+    // inbound group sessions. We don't populate it yet; that happens once we
+    // have done the upgrade to v6, in `migrate_data_for_v6`. Finally we drop the
+    // old store in create_stores_for_v7.
+
+    info!("Creating {}", keys::INBOUND_GROUP_SESSIONS);
+    let object_store = db.create_object_store(keys::INBOUND_GROUP_SESSIONS)?;
+
+    let mut params = IdbIndexParameters::new();
+    params.unique(false);
+    object_store.create_index_with_params(
+        keys::INBOUND_GROUP_SESSIONS_BACKUP_INDEX,
+        &IdbKeyPath::str("needs_backup"),
+        &params,
+    )?;
+
+    Ok(())
+}
+
+async fn migrate_data_for_v6(serializer: &IndexeddbSerializer, db: &IdbDatabase) -> Result<()> {
+    // The new store has been made for inbound group sessions; time to populate it.
+    let txn = db.transaction_on_multi_with_mode(
+        &[old_keys::INBOUND_GROUP_SESSIONS_OLD, keys::INBOUND_GROUP_SESSIONS],
+        IdbTransactionMode::Readwrite,
+    )?;
+
+    let old_store = txn.object_store(old_keys::INBOUND_GROUP_SESSIONS_OLD)?;
+    let new_store = txn.object_store(keys::INBOUND_GROUP_SESSIONS)?;
+
+    let row_count = old_store.count()?.await?;
+    info!(row_count, "Migrating inbound group session data");
+
+    if let Some(cursor) = old_store.open_cursor()?.await? {
+        let mut idx = 0;
+        loop {
+            idx += 1;
+            let key = cursor.key().ok_or(matrix_sdk_crypto::CryptoStoreError::Backend(
+                "inbound_group_sessions cursor has no key".into(),
+            ))?;
+            let value = cursor.value();
+
+            if idx % 100 == 0 {
+                debug!("Migrating session {idx} of {row_count}");
```

Comment:
Any idea how many log lines would this be for a super-busy power user account?

Reply:
Not *really*, tbh. I have 17000 rows in my store, which means 170 lines.

Something worth emphasising here: there are currently about 16 users of Element R. Once those 16 users have done this migration, we'll never hit this line of code again. As such, it's probably not worth worrying *too* much about it.